conan/tools/cmake/toolchain.py

```diff
@@ -394,28 +395,84 @@ def context(self):
         return ctxt_toolchain
 
 
-class FindConfigFiles(Block):
+class FindFiles(Block):
     template = textwrap.dedent("""
         {% if find_package_prefer_config %}
         set(CMAKE_FIND_PACKAGE_PREFER_CONFIG {{ find_package_prefer_config }})
         {% endif %}
-        # To support the generators based on find_package()
-        {% if cmake_module_path %}
-        set(CMAKE_MODULE_PATH {{ cmake_module_path }} ${CMAKE_MODULE_PATH})
+        {% if generators_folder or host_build_paths_root or host_build_paths_noroot or build_build_paths %}
+        set(CMAKE_MODULE_PATH {{ generators_folder }} {{ host_build_paths_root }} {{ host_build_paths_noroot }} {{ build_build_paths }} ${CMAKE_MODULE_PATH})
         {% endif %}
-        {% if cmake_prefix_path %}
-        set(CMAKE_PREFIX_PATH {{ cmake_prefix_path }} ${CMAKE_PREFIX_PATH})
+        {% if generators_folder or host_build_paths_noroot %}
+        set(CMAKE_PREFIX_PATH {{ generators_folder }} {{ host_build_paths_noroot }} ${CMAKE_PREFIX_PATH})
         {% endif %}
-        {% if android_prefix_path %}
-        set(CMAKE_FIND_ROOT_PATH {{ android_prefix_path }} ${CMAKE_FIND_ROOT_PATH})
+        {% if cmake_program_path %}
+        set(CMAKE_PROGRAM_PATH {{ cmake_program_path }} ${CMAKE_PROGRAM_PATH})
+        {% endif %}
+        {% if cmake_library_path %}
+        set(CMAKE_LIBRARY_PATH {{ cmake_library_path }} ${CMAKE_LIBRARY_PATH})
+        {% endif %}
+        {% if is_apple and cmake_framework_path %}
+        set(CMAKE_FRAMEWORK_PATH {{ cmake_framework_path }} ${CMAKE_FRAMEWORK_PATH})
+        {% endif %}
+        {% if cmake_include_path %}
+        set(CMAKE_INCLUDE_PATH {{ cmake_include_path }} ${CMAKE_INCLUDE_PATH})
         {% endif %}
 
-        # To support cross building to iOS, watchOS and tvOS where CMake looks for config files
-        # only in the system frameworks unless you declare the XXX_DIR variables
-        {% if cross_ios %}
-            set(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE BOTH)
+        {% if cross_building %}
+        if(NOT DEFINED CMAKE_FIND_ROOT_PATH_MODE_PACKAGE)
+            set(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE "BOTH")
+        endif()
+        if(NOT DEFINED CMAKE_FIND_ROOT_PATH_MODE_PROGRAM)
+            set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM "BOTH")
+        endif()
+        if(NOT DEFINED CMAKE_FIND_ROOT_PATH_MODE_LIBRARY)
+            set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY "BOTH")
+        endif()
+        {% if is_apple %}
+        if(NOT DEFINED CMAKE_FIND_ROOT_PATH_MODE_FRAMEWORK)
+            set(CMAKE_FIND_ROOT_PATH_MODE_FRAMEWORK "BOTH")
+        endif()
         {% endif %}
-        """)
+        if(NOT DEFINED CMAKE_FIND_ROOT_PATH_MODE_INCLUDE)
+            set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE "BOTH")
+        endif()
+
+        if(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE STREQUAL "ONLY")
+            set(CMAKE_FIND_ROOT_PATH ${CMAKE_FIND_ROOT_PATH} {{ generators_folder }} {{ host_build_paths_noroot }})
```

Comment:
is this trick necessary? why not set `BOTH`? I'm sure there is an explanation.

Reply:
Sure, the toolchain could force `CMAKE_FIND_ROOT_PATH_MODE_PACKAGE` to BOTH. But I would like to avoid to override this variable if set by user toolchain.

Reply:
I'm wondering, if the user toolchain is setting `CMAKE_FIND_ROOT_PATH_MODE_PACKAGE STREQUAL "ONLY"`, wouldn't be mostly the same playing the `CMAKE_FIND_ROOT_PATH`. Why is that ok but not override the variable?

Reply:
Not sure to understand the questions. Would you like to set `CMAKE_FIND_ROOT_PATH` with generator folder and non-root builddirs of requires, regardless of `CMAKE_FIND_ROOT_PATH_MODE_PACKAGE` value?

Reply:
No, just saying that you are protecting the user set value from `CMAKE_FIND_ROOT_PATH_MODE_PACKAGE` because `I would like to avoid to override this variable if set by user toolchain.` but then you are modifying the `CMAKE_FIND_ROOT_PATH`, so I'm wondering if it is not kind of "cheating".

Reply:
In other words, why is so important to keep the user value of `CMAKE_FIND_ROOT_PATH_MODE_PACKAGE` (and others)?

Reply:
Ok, so when `CMAKE_FIND_ROOT_PATH_MODE_*` is `ONLY`, you have no other choice than manipulating `CMAKE_FIND_ROOT_PATH` because all other variables are ignored. `CMAKE_FIND_ROOT_PATH` is like `CMAKE_PREFIX_PATH` but with higher precedence, so no it's not cheating.

Why is it important to honor `CMAKE_FIND_ROOT_PATH_MODE_*` of users? Principle of least surprise.
But as I said before, if `ONLY` is overridden by `BOTH`, it should not break consumers (other proposal in https://github.com/conan-io/conan/pull/10186#discussion_r774247442), because system paths have the lowest priority in search paths of `find_*()` commands, and `BOTH` still look into `CMAKE_FIND_ROOT_PATH` first if the user has injected some stuff to managed by conan.

Reply:
Then I would simplify and just override to `BOTH`. 

> Ok, so when CMAKE_FIND_ROOT_PATH_MODE_* is ONLY, you have no other choice than manipulating CMAKE_FIND_ROOT_PATH because all other variables are ignored. CMAKE_FIND_ROOT_PATH is like CMAKE_PREFIX_PATH but with higher precedence, so no it's not cheating.

Yes, I got that, but:

> Why is it important to honor CMAKE_FIND_ROOT_PATH_MODE_* of users? Principle of least surprise.

Not sure that changing `CMAKE_FIND_ROOT_PATH` is `least surprise` either, so I vote for simplifying and just overriding to BOTH when cross_build.

Reply:
simplified in https://github.com/conan-io/conan/pull/10186/commits/1e89fe920cc491611dbff41b377e29100e323219

`CMAKE_FIND_ROOT_PATH_MODE_*` of user honored unless it's `ONLY`.